File: 46-permutations/permutations.cpp
class Solution {
public:
    vector<vector<int>> result;

    void backtrack(vector<int>& nums, int index) {
        // If all positions are fixed
        if (index == nums.size()) {
            result.push_back(nums);
            return;
        }

        for (int i = index; i < nums.size(); i++) {
            swap(nums[index], nums[i]);          // choose
            backtrack(nums, index + 1);          // recurse
            swap(nums[index], nums[i]);          // backtrack
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        backtrack(nums, 0);
        return result;
    }
};

File: 46-permutations/permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []

        def backtrack(path, remaining):
            # If no elements left, store permutation
            if not remaining:
                result.append(path)
                return

            # Try each number
            for i in range(len(remaining)):
                backtrack(
                    path + [remaining[i]],                 # add current number
                    remaining[:i] + remaining[i+1:]       # remaining numbers
                )

        backtrack([], nums)
        return result